### PR TITLE
Set names

### PR DESCRIPTION
**Pull request by martinsvalin**

### CODE REVIEW

File: assets/js/app.js [hunk 1]
```diff
@@ -18,6 +18,14 @@ import NProgress from "nprogress"
 import { LiveSocket } from "phoenix_live_view"
 import Hooks from "./hooks"
 
+window.rememberName = () => {
+    let name = document.getElementById("name").value
+    let xhr = new XMLHttpRequest()
+    xhr.open("POST", window.location.origin + "/__api/set_name", true)
+    xhr.setRequestHeader("Content-Type", "application/json")
+    xhr.send(JSON.stringify({ name: name }))
+}
```
Comment (marreman) [inline assets/js/app.js:27]: Could this be done in a hook instead?

File: assets/js/app.js [hunk 2]
```diff
@@ -18,6 +18,14 @@ import NProgress from "nprogress"
 import { LiveSocket } from "phoenix_live_view"
 import Hooks from "./hooks"
 
+window.rememberName = () => {
+    let name = document.getElementById("name").value
+    let xhr = new XMLHttpRequest()
```
Comment (marreman) [inline assets/js/app.js:23]: We can use the more modern fetch API here. 

https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

File: lib/spyfall_web/live/game_live.ex [hunk 3]
```diff
@@ -32,6 +35,11 @@ defmodule SpyfallWeb.GameLive do
     {:noreply, socket}
   end
 
+  def handle_event("new_name", %{"value" => new_name}, socket) do
+    Player.update(self(), "game:" <> socket.assigns.game, socket.id, %{name: new_name})
+    {:noreply, socket}
```
Comment (marreman) [inline lib/spyfall_web/live/game_live.ex:40]: Can we assign the name to the session here instead?

File: assets/js/app.js (hunk 2 above)
Comment (martinsvalin) [reply assets/js/app.js:23]: See, this is why I PRed this :) old habits

File: assets/js/app.js (hunk 1 above)
Comment (martinsvalin) [reply assets/js/app.js:27]: I think hooks are generally meant for the client to react when the server does something. Here, it’s the client doing a thing. I guess we could do it with a hook, though, reacting to the update coming back when we change the Presence state. 
Or maybe I misinterpreted hooks, and they can be used to hook in on outgoing client events.

Comment (martinsvalin) [reply assets/js/app.js:27]: I think hooks are generally meant for the client to react when the server does something. Here, it’s the client doing a thing. I guess we could do it with a hook, though, reacting to the update coming back when we change the Presence state. 
Or maybe I misinterpreted hooks, and they can be used to hook in on outgoing client events.

File: lib/spyfall_web/live/game_live.ex (hunk 3 above)
Comment (martinsvalin) [reply lib/spyfall_web/live/game_live.ex:40]: It _would_ seem like the natural place, right? But we’re acting on websocket data and only have access to the data sent over the wire and the socket process state. We can’t see the session here, and there are no cookies. 

It made me start thinking of a proposal to LiveView. We do have the session on mount. We could keep it in the socket process state (socket.session), and have a specific api mimicking the regular `Plug.Conn.put_session/3` but working with a liveview socket. Storing something in the session would have the side effect of talking to the client side liveview library to set a new session in the cookie client side. 

Feels doable, but also feels a bit like we’re reimplementing http features we’re missing in web sockets. There can be weird failure states where client and server have different ideas of what’s in the session if that call fails.

Comment (martinsvalin) [reply lib/spyfall_web/live/game_live.ex:40]: It _would_ seem like the natural place, right? But we’re acting on websocket data and only have access to the data sent over the wire and the socket process state. We can’t see the session here, and there are no cookies. 

It made me start thinking of a proposal to LiveView. We do have the session on mount. We could keep it in the socket process state (socket.session), and have a specific api mimicking the regular `Plug.Conn.put_session/3` but working with a liveview socket. Storing something in the session would have the side effect of talking to the client side liveview library to set a new session in the cookie client side. 

Feels doable, but also feels a bit like we’re reimplementing http features we’re missing in web sockets. There can be weird failure states where client and server have different ideas of what’s in the session if that call fails.

Comment (martinsvalin) [reply lib/spyfall_web/live/game_live.ex:40]: It _would_ seem like the natural place, right? But we’re acting on websocket data and only have access to the data sent over the wire and the socket process state. We can’t see the session here, and there are no cookies. 

It made me start thinking of a proposal to LiveView. We do have the session on mount. We could keep it in the socket process state (socket.session), and have a specific api mimicking the regular `Plug.Conn.put_session/3` but working with a liveview socket. Storing something in the session would have the side effect of talking to the client side liveview library to set a new session in the cookie client side. 

Feels doable, but also feels a bit like we’re reimplementing http features we’re missing in web sockets. There can be weird failure states where client and server have different ideas of what’s in the session if that call fails.

File: assets/js/app.js (hunk 2 above)
Comment (henrik) [reply assets/js/app.js:23]: Hehe, yes, doing this in my version of this… blogging it as we speak :D 

```
fetch(`/api/session?$foo=${encodeURIComponent(bar)}`, { method: "post" })
```

File: assets/js/app.js (hunk 1 above)
Comment (martinsvalin) [reply assets/js/app.js:27]: I tried out setting up an event listened on the mounted hook. Feels reasonable, so I'll go with this.

**Comment by henrik** [comment]
You may want to also increase the lifetime of the session cookie.

Just blogged my take on this here – let me know if you'd like me to link to this repo/code specifically, in the credits. Wasn't sure if you'd want me to :)

https://thepugautomatic.com/2020/05/persistent-session-data-in-phoenix-liveview/

File: assets/js/app.js (hunk 1 above)
Comment (henrik) [reply assets/js/app.js]: Yeah, I believe I've seen "official" code use hooks that way 👍

File: lib/spyfall_web/live/game_live.ex (hunk 3 above)
Comment (marreman) [reply lib/spyfall_web/live/game_live.ex:40]: Wouldn’t it be nice to just have it as an argument to the handle_event function. Like in mount?

Comment (marreman) [reply lib/spyfall_web/live/game_live.ex:40]: Or maybe live view doesn’t send the session cookie ID with every request. Probably not.

**Comment by marreman** [comment]
Maybe it’s better to put the name in local storage than in the session?

**Pull request merged**